File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/2. Interview Problems/python/01. Minimum bit flips.py
"""QUESTION:
Given two integers start and goal, return the minimum number of bit flips required to convert start to goal.

Approach:
Count the positions where the bits differ by shifting both numbers right and comparing the least-significant bits.

Time Complexity: O(1) (32 iterations)
Space Complexity: O(1)
"""

def minBitFlips(start: int, goal: int) -> int:
    i = 0
    cnt = 0
    while i < 32:
        start_bit = (start >> i) & 1
        goal_bit = (goal >> i) & 1
        if start_bit != goal_bit:
            cnt += 1
        i += 1
    return cnt

File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/2. Interview Problems/javascript/01. Minimum bit flips.js
/**
QUESTION:
Given two integers `start` and `goal`, return the minimum number of bit flips required to convert `start` to `goal`.

Approach:
Count the positions where the bits differ. Shift both numbers right and compare the least-significant bits until all 32 positions are processed.

Time Complexity: O(1) (32 iterations)
Space Complexity: O(1)
*/

function minBitFlips(start, goal) {
    let i = 0;
    let cnt = 0;
    while (i < 32) {
        const startBit = (start >> i) & 1;
        const goalBit = (goal >> i) & 1;
        if (startBit !== goalBit) cnt++;
        i++;
    }
    return cnt;
}

module.exports = { minBitFlips };
